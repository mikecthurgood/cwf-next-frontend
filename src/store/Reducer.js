const reducer = (state, action) => {
    switch (action.type) {
        
        case 'setFilterSelection':
            return { ...state, filterSelection: action.data}
        case 'setLoginError':
            return { ...state, loginError: action.data}
        case 'setLoginMenuVisible':
            return { ...state, loginMenuVisible: action.data}
        case 'setMobileMenuVisible':
            return { ...state, mobileMenuVisible: action.data}
        case 'setSearchBarVisibility':
            return { ...state, searchBarVisible: action.data}
        case 'setSearchFilter':
            return { ...state, searchFilter: action.data}
        case 'setScrollPosition':
            return { ...state, scrollPosition: action.data}
        case 'setSignUpSuccess':
            return { ...state, signUpSuccess: action.data}
        case 'setSingleWall':
            return { ...state, singleWall: action.data}
        case 'setSortInputVisibility':
            return { ...state, sortInputVisible: action.data}
        case 'setUser':
            return { ...state, user: action.data}
        case 'setPostCode':
            return { ...state, userPostCode: action.data}
        case 'setWalls':
            return { ...state, walls: action.data}
        case 'signOut' :
            localStorage.removeItem('token')
            localStorage.removeItem('userName')
            localStorage.removeItem('userId')
            return { 
                ...state,
                user: {username: null, userId: null, isAuth: false, token: ''},
                mobileMenuVisible: false,
                loginMenuVisible: false
            }
        default: 
            throw new Error('Invalid action type dispatched')
    }
}
export default reducer