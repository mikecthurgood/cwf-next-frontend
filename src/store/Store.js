import React, { useReducer } from  'react';
import reducer from './Reducer'

const initialState = {
    filterSelection: [],
    loginError: false,
    loginMenuVisible: false,
    mobileMenuVisible: false,
    searchBarVisible: false,
    scrollPosition: 0,
    signUpSuccess: false,
    singleWall: {},
    sortInputVisible: false,
    user: { username: null, userId: null, isAuth: false, token: '' },
    userPostCode: '',
    walls: [],
}

const State = React.createContext(initialState)
const Dispatch = React.createContext(null)

const Store = ({children}) => {
    
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <Dispatch.Provider value={dispatch} >
            <State.Provider value={state} >
                {children}
            </State.Provider>
        </Dispatch.Provider>
    )
}


export { State, Dispatch, Store }

