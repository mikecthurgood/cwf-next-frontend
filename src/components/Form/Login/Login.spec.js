import React from 'react';
import '@testing-library/jest-dom';
import { cleanup, render, fireEvent, waitFor } from '@testing-library/react';
import LoginForm from '.';
import { Dispatch, State } from '../../../store/Store';

jest.mock('../../../services/helpers/AuthHelpers');

describe('Login Form', () => {
    afterEach(() => {
        cleanup()
    })
    const dispatch = jest.fn();
    const state = {
        loginError: false
    };
    it('renders the component', () => {
        
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm />
                </State.Provider>
            </Dispatch.Provider>
        );
        
        expect(container).toMatchSnapshot();
    })

    it('renders the component in compact view', () => {
        
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm layout={'compact'} />
                </State.Provider>
            </Dispatch.Provider>
        );
        
        expect(container).toMatchSnapshot();
    })

    it('allows a username and passwors to be input', () => {
        
        const {getByPlaceholderText} = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm />
                </State.Provider>
            </Dispatch.Provider>
        );
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');

        fireEvent.change(emailInput, { target: { value: 'test@test.com' } })
        fireEvent.change(passwordInput, { target: { value: 'qwerty' } })
        expect(emailInput.value).toEqual('test@test.com');
        expect(passwordInput.value).toEqual('qwerty');
    })

    it('allows a username and password to be input', () => {
        
        const {getByPlaceholderText, getByText} = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm />
                </State.Provider>
            </Dispatch.Provider>
        );
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');
        const loginButton = getByText('Login');
        fireEvent.change(emailInput, { target: { value: 'test@test.com' } })
        fireEvent.change(passwordInput, { target: { value: 'qwerty' } })
        expect(loginButton.closest('button')).not.toBeDisabled();
    })

    it('submits a login event', () => {
        const {getByPlaceholderText, getByText} = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm />
                </State.Provider>
            </Dispatch.Provider>
        );
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');
        const loginButton = getByText('Login');
        fireEvent.change(emailInput, { target: { value: 'test@test.com' } })
        fireEvent.change(passwordInput, { target: { value: 'qwerty' } })
        fireEvent.click(loginButton);
        expect(dispatch).toBeCalledWith({
            type: 'loginHandlerDispatchTest',
            data: {
                submitData: {
                    email: 'test@test.com',
                    password: 'qwerty'
                }
            },
        })
    })

    it('sets a login error if login is unsuccessful', () => {
        const {getByPlaceholderText, getByText} = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <LoginForm />
                </State.Provider>
            </Dispatch.Provider>
        );
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');
        const loginButton = getByText('Login');
        fireEvent.change(emailInput, { target: { value: 'failme' } })
        fireEvent.change(passwordInput, { target: { value: 'qwerty' } })
        fireEvent.click(loginButton);
        waitFor(() => {
            expect(dispatch).toBeCalledWith({
                type: 'setLoginError',
                data: true,
            })
        })
    })
})