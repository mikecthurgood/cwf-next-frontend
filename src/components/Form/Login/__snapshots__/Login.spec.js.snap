// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders the component 1`] = `
<div>
  <div
    class="Login__AuthFormContainer-sc-1ryrs0l-0 hBRKZY"
  >
    <form
      class="Login__Form-sc-1ryrs0l-1 SwwBW"
    >
      <button
        class="Login__AuthButton-sc-1ryrs0l-3 izDnDC"
      >
        Register
      </button>
      <div
        class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
      >
        <div
          class="Input__InputContainer-sc-1wyp7tv-0 kIYPqy"
        >
          <input
            class="invalid untouched "
            data-testid="form-input"
            id="userEmail"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
      >
        <div
          class="Input__InputContainer-sc-1wyp7tv-0 kIYPqy"
        >
          <input
            class="invalid untouched "
            data-testid="form-input"
            id="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="Login__AuthButton-sc-1ryrs0l-3 izDnDC disabled"
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`Login Form renders the component in compact view 1`] = `
<div>
  <div
    class="Login__AuthFormContainer-sc-1ryrs0l-0 hBRKZY"
  >
    <form
      class="Login__Form-sc-1ryrs0l-1 NjHaG"
    >
      <div
        class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
      >
        <div
          class="Input__InputContainer-sc-1wyp7tv-0 eyXXuZ"
        >
          <input
            class="invalid untouched "
            data-testid="form-input"
            id="userEmail"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="Login__FormInput-sc-1ryrs0l-2 iNdDlt"
      >
        <div
          class="Input__InputContainer-sc-1wyp7tv-0 eyXXuZ"
        >
          <input
            class="invalid untouched "
            data-testid="form-input"
            id="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="Login__AuthButton-sc-1ryrs0l-3 emqpgl disabled"
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
